hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeHandler.java

```diff
@@ -72,51 +110,96 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+                    freezeTxn, txValidStart);
 
-                // FREEZE_ABORT does not require any additional checks
-            case FREEZE_ABORT -> {
+                // PREPARE_UPGRADE, FREEZE_ABORT do not require any additional checks
+            default -> {
                 // do nothing
             }
         }
+    }
 
-        // no need to add any keys to the context because this transaction does not require any signatures
-        // it must be submitted by an account with superuser privileges, that is checked during ingest
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull final NetworkAdminServiceConfig adminServiceConfig,
+            @NonNull final ReadableSpecialFileStore specialFileStore,
+            @NonNull final WritableFreezeStore freezeStore) {
+        requireNonNull(txn);
+        requireNonNull(adminServiceConfig);
+        requireNonNull(specialFileStore);
+        requireNonNull(freezeStore);
+        final FreezeTransactionBody freezeTxn = txn.freezeOrThrow();
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
+
+        validateSemantics(freezeTxn, specialFileStore, updateFileNum);
+
+        final FreezeUpgradeActions upgradeActions = new FreezeUpgradeActions(adminServiceConfig, freezeStore);
+        final Timestamp freezeStartTime = freezeTxn.startTime(); // may be null for some freeze types
+        final Instant freezeStartTimeInstant = freezeStartTime == null
+                ? null
+                : Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+
+        switch (freezeTxn.freezeType()) {
+            case PREPARE_UPGRADE ->
+            // by the time we get here, we've already checked that updateFileNum is non-null in validateSemantics()
+            upgradeActions.extractSoftwareUpgrade(specialFileStore
+                    .get(requireNonNull(updateFileNum).fileNum())
+                    .orElseThrow(() -> new IllegalStateException("Update file not found")));
+                // TODO: call networkCtx.recordPreparedUpgrade(freezeTxn); (issue #6201)
+            case FREEZE_UPGRADE -> upgradeActions.scheduleFreezeUpgradeAt(requireNonNull(freezeStartTimeInstant));
+            case FREEZE_ABORT -> upgradeActions.abortScheduledFreeze();
+                // TODO: call networkCtx.discardPreparedUpgradeMeta(); (issue #6201)
+            case TELEMETRY_UPGRADE -> upgradeActions.extractTelemetryUpgrade(
+                    specialFileStore
+                            .get(requireNonNull(updateFileNum).fileNum())
+                            .orElseThrow(() -> new IllegalStateException("Telemetry update file not found")),
+                    requireNonNull(freezeStartTimeInstant));
+                // case FREEZE_ONLY is default
+            default -> upgradeActions.scheduleFreezeOnlyAt(requireNonNull(freezeStartTimeInstant));
```

Comment:
If you include every enumerated value as a case, and do *not* include a default, then the compiler will tell us if we forget to update this in the future (and fail to compile).

Reply:
Done.